### PR TITLE
dialects: (arith) split generic binary op definition into specific ones

### PR DESCRIPTION
**Pull request by superlopuh**
This makes the next step for Pyright updating easier, as we can't actually properly represent generics in the IRDL constraint system. Currently, the arith base class helpers work because the generics are all specified by the time the ops are "defined" with the annotation. But my proposed solution of replacing the Annotated with VarConstraints means that we need to define the constraint on a specific type already. My understanding is that this is not a functional change, only requiring some clients of the base classes to update the names that they refer to.

Part of #3264

Note stacked PR.

### CODE REVIEW

File: xdsl/dialects/arith.py [hunk 1]
```diff
@@ -214,10 +213,15 @@ def __hash__(self) -> int:
         return id(self)
 
 
-SignlessIntegerBinaryOp = BinaryOperation[Annotated[Attribute, signlessIntegerLike]]
+class FloatingPointLikeBinaryOperation(IRDLOperation, abc.ABC):
```
Comment (superlopuh) [inline xdsl/dialects/arith.py:216]: turns out all the floating ops below are actually with fastmath

File: xdsl/dialects/arith.py [hunk 2]
```diff
@@ -255,13 +264,6 @@ def print(self, printer: Printer):
         printer.print_attribute(self.result.type)
 
 
-FloatingPointLikeBinaryOp = BinaryOperationWithFastMath[
-    Annotated[Attribute, floatingPointLike]
-]
-
-IntegerBinaryOp = BinaryOperation[IntegerType]
```
Comment (superlopuh) [inline xdsl/dialects/arith.py:262]: this is not used

File: tests/irdl/test_declarative_assembly_format.py [hunk 3]
```diff
@@ -1661,6 +1664,18 @@ class ParamOne(ParametrizedAttribute, TypeAttribute, Generic[_T]):
...
+            cls,
+            *,
+            n: GenericAttrConstraint[Attribute] | None = None,
+            p: GenericAttrConstraint[_T] | None = None,
+            q: GenericAttrConstraint[Attribute] | None = None,
+        ) -> BaseAttr[ParamOne[Attribute]] | ParamAttrConstraint[ParamOne[_T]]:
+            if n is None and p is None and q is None:
+                return BaseAttr(cls)
+            return ParamAttrConstraint(cls, (n, p, q))
+
```
Comment (math-fehr) [inline tests/irdl/test_declarative_assembly_format.py]: Are those changes related?

File: tests/tblgen_to_py/test.py [hunk 4]
```diff
@@ -50,13 +50,7 @@ class Test_AttributesOp(IRDLOperation):
...
 
     int_attr = prop_def(
-        ParamAttrConstraint(
-            IntegerAttr,
-            (
-                AnyAttr(),
-                EqAttrConstraint(IntegerType(16)),
-            ),
-        )
+        IntegerAttr[IntegerType].constr(type=EqAttrConstraint(IntegerType(16)))
```
Comment (math-fehr) [inline tests/tblgen_to_py/test.py]: Same in this file

File: xdsl/dialects/arith.py [hunk 5]
```diff
@@ -171,12 +172,10 @@ def parse(cls: type[Constant], parser: Parser) -> Constant:
 _T = TypeVar("_T", bound=Attribute)
 
 
-class BinaryOperation(IRDLOperation, Generic[_T]):
-    """A generic base class for arith's binary operation.
+class SignlessIntegerBinaryOperation(IRDLOperation, abc.ABC):
+    """A generic base class for arith's binary operation on signless integers."""
```
Comment (alexarice) [inline xdsl/dialects/arith.py:176]:
```suggestion
    """A generic base class for arith's binary operations on signless integers."""
```

**Review by math-fehr — APPROVED** [review]
🚀 
Sad to see it gone, but it doesn't matter that much I guess

File: xdsl/dialects/arith.py [hunk 6]
```diff
@@ -271,7 +273,7 @@ def get_canonicalization_patterns(cls) -> tuple[RewritePattern, ...]:
 
 
 @irdl_op_definition
-class Addi(SignlessIntegerBinaryOp):
+class Addi(SignlessIntegerBinaryOperation):
```
Comment (alexarice) [inline xdsl/dialects/arith.py:276]: Any particular reason to change the name here?

File: xdsl/transforms/varith_transformations.py [hunk 7]
```diff
@@ -15,7 +15,8 @@
 
 # map the arith operation to the right varith op:
 ARITH_TO_VARITH_TYPE_MAP: dict[
-    type[arith.BinaryOperation[Attribute]], type[varith.VarithOp]
+    type[arith.SignlessIntegerBinaryOperation | arith.FloatingPointLikeBinaryOperation],
```
Comment (alexarice) [inline xdsl/transforms/varith_transformations.py:18]: it is worth making a type alias for this?

**Review by alexarice — APPROVED** [review]
Looks good

**Comment by alexarice** [comment]
I can't even see the changes Mathieu reviewed, was this just a merging/rebasing mishap?

File: xdsl/dialects/arith.py (hunk 6 above)
Comment (superlopuh) [reply xdsl/dialects/arith.py:276]: The undocumented recommendation is for ABCs to have the full Operation and concrete ops to have Op as a suffix

File: xdsl/transforms/varith_transformations.py (hunk 7 above)
Comment (superlopuh) [reply xdsl/transforms/varith_transformations.py:18]: IDK, I wouldn't bother

File: xdsl/dialects/builtin.py [hunk 8]
```diff
@@ -511,6 +513,24 @@ def parse_with_type(
...
+        if value is None and type == AnyAttr():
+            return BaseAttr[IntegerAttr[_IntegerAttrType]](IntegerAttr)
+        return ParamAttrConstraint[IntegerAttr[_IntegerAttrType]](
+            IntegerAttr,
+            (
+                value,
+                type,
+            ),
+        )
+
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/dialects/builtin.py]: 🧙 🪄

File: tests/tblgen_to_py/test.py (hunk 4 above)
Comment (superlopuh) [reply tests/tblgen_to_py/test.py]: 🧙 🪄

File: xdsl/dialects/arith.py [hunk 9]
```diff
@@ -214,10 +213,15 @@ def __hash__(self) -> int:
         return id(self)
 
 
-SignlessIntegerBinaryOp = BinaryOperation[Annotated[Attribute, signlessIntegerLike]]
+class FloatingPointLikeBinaryOperation(IRDLOperation, abc.ABC):
+    """A generic base class for arith's binary operation on floats."""
```
Comment (superlopuh) [inline xdsl/dialects/arith.py:217]:
```suggestion
    """A generic base class for arith's binary operations on floats."""
```

**Review by alexarice — APPROVED** [review]

**Pull request merged**